### PR TITLE
Add support for Empty Sector Update proofs (for SnapDeals)

### PR DESCRIPTION
**Pull request by cryptonemo**

### CODE REVIEW

**Comment by cryptonemo** [comment]
Important note:  while all tests are passing locally, I have verified that the `empty_sector_update_circuit*` tests are failing on CI because they top out around 42GB RAM, and so they are killed in the CI environment.  This is an issue that can be addressed separately, as it does not affect production immediately and is otherwise acceptable for initial integration work.

File: storage-proofs-update/src/vanilla.rs [hunk 1]
```diff
@@ -0,0 +1,1209 @@
...
+            .try_for_each(|(chunk_index, replica_data)| -> Result<()> {
+                for i in (0..data_block_size as u64).step_by(FR_SIZE) {
+                    let input_index = (chunk_index as usize) + i as usize;
+                    let output_index = i as usize;
+
+                    // Get the `h` high bits from the node-index (sans partition-index).
+                    let node_index = input_index / FR_SIZE;
+                    let input_index_without_k = node_index & remove_k_from_node_index_mask;
+                    let high =
+                        input_index_without_k >> (node_index_bit_len - partition_bit_len - h);
```
Comment (Kubuxu) [inline storage-proofs-update/src/vanilla.rs:892]: I don't think `parititon_bit_len` should be used here.
We want `h` high bits out of the node index, the partition count is unimportant to the encoding process.

File: storage-proofs-update/src/gadgets.rs [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+    assert_eq!(h_select_bits.len(), hs.len());
+
+    let bit_len = c_generated_bits.len();
+
+    let c_generated_bits: Vec<Boolean> = c_generated_bits.iter().cloned().map(Into::into).collect();
+
+    // For each `h in hs`, get the `h` high bits of the challenge's randomly generated bits (i.e.
+    // the challenge's bits sans the partition-index). Scale each "high" value by the corresponding
+    // bit of `h_select` producing a vector containing `hs.len() - 1` zeros and 1 "high" value
+    // selected via `h_select_bits`.
```
Comment (Kubuxu) [inline storage-proofs-update/src/gadgets.rs:307]: `(i.e. the challenge's bits sans the partition-index).` this part of the comment is wrong, I think it got out of date.

File: storage-proofs-update/src/circuit.rs [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+            let c_bits: Vec<AllocatedBit> = c_bits_without_partition
+                .iter()
+                .chain(partition_bits.iter())
+                .cloned()
+                .collect();
+
+            // Compute this challenge's `rho`.
+            let c_high = get_challenge_high_bits(
+                cs.namespace(|| format!("get_challenge_high_bits (c_index={})", c_index)),
+                &c_bits_without_partition,
```
Comment (Kubuxu) [inline storage-proofs-update/src/circuit.rs:433]: c_high for `rho` generation should include partition bits, I think we misscommunicated.

File: storage-proofs-update/src/vanilla.rs [hunk 4]
```diff
@@ -0,0 +1,1209 @@
...
+                        &Fr::from(high as u64).into(),
+                    )
+                    .into();
+
+                    let sector_key_fr =
+                        bytes_into_fr(&sector_key_data[input_index..input_index + FR_SIZE])?;
+                    let replica_data_fr =
+                        bytes_into_fr(&replica_data[input_index..input_index + FR_SIZE])?;
+
+                    let out_data_fr = (replica_data_fr - sector_key_fr) * rho.invert().unwrap();
```
Comment (Kubuxu) [inline storage-proofs-update/src/vanilla.rs:1068]: Same here, we don't want to compute \rho and invert it on a per-node basis.

File: filecoin-proofs/src/api/update.rs [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+#[allow(clippy::too_many_arguments)]
+pub fn encode_into<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
+    porep_config: PoRepConfig,
+    new_replica_path: &Path,
+    new_cache_path: &Path,
+    sector_key_path: &Path,
+    sector_key_cache_path: &Path,
+    staged_data_path: &Path,
+    piece_infos: &[PieceInfo],
+) -> Result<(Commitment, Commitment, Commitment)> {
```
Comment (dignifiedquire) [inline filecoin-proofs/src/api/update.rs:86]: Please use a struct for this return value, super easy to misuse the API otherwise

File: filecoin-proofs/src/types/hselect.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#[derive(Clone, Copy, Debug)]
+pub struct HSelect(pub u8);
```
Comment (dignifiedquire) [inline filecoin-proofs/src/types/hselect.rs:2]: please document what this is/does

File: filecoin-proofs/src/types/update_proof_partitions.rs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#[derive(Clone, Copy, Debug)]
+pub struct UpdateProofPartitions(pub u8);
```
Comment (dignifiedquire) [inline filecoin-proofs/src/types/update_proof_partitions.rs:2]: please document

File: storage-proofs-update/src/circuit.rs [hunk 8]
```diff
@@ -0,0 +1,587 @@
...
+            path_r_old,
+            leaf_d_new,
+            path_d_new,
+            leaf_r_new,
+            path_r_new,
+            _tree_r: PhantomData,
+        }
+    }
+
+    pub fn blank(sector_nodes: usize) -> Self {
```
Comment (dignifiedquire) [inline storage-proofs-update/src/circuit.rs:135]: I think we use `empty`  in the other circuits, can we keep consistent with that?

File: storage-proofs-core/src/gadgets/por.rs [hunk 9]
```diff
@@ -505,7 +505,13 @@ where
         )?;
     }
     let computed_root = cur;
-    constraint::equal(cs, || "merkle root equality", &computed_root, &root);
```
[the remark this answers is not in the document]
Comment (Kubuxu) [reply storage-proofs-core/src/gadgets/por.rs:508]: I'm under the impression that the `por_no_challenge_input` gadget was not used anywhere but it needs to be verified.

File: storage-proofs-update/src/vanilla.rs [hunk 10]
```diff
@@ -0,0 +1,1183 @@
...
+        comm_c: TreeRDomain,
+        comm_r_last_old: TreeRDomain,
+        new_replica_path: &Path,
+        new_cache_path: &Path,
+        sector_key_path: &Path,
+        sector_key_cache_path: &Path,
+        staged_data_path: &Path,
+        h: usize,
+    ) -> Result<(TreeRDomain, TreeRDomain, TreeDDomain)> {
+        // Sanity check all input path types.
```
Comment (qy3u) [inline storage-proofs-update/src/vanilla.rs:784]: May be also ensure that the `new_replica_path` (and cache path) is not equal to the old one here?

File: storage-proofs-update/src/circuit.rs (hunk 8 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/circuit.rs:135]: Yup, I can do that. I chose `blank` because its called by `EmptySectorUpdateCompound::blank_circuit`.

File: storage-proofs-core/src/gadgets/por.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-core/src/gadgets/por.rs:508]: Or `storage_proofs_update::gadgets::por_no_challenge_input` should be moved into `storage_proofs_core::gadgets`.

File: storage-proofs-update/src/vanilla.rs [hunk 11]
```diff
@@ -0,0 +1,1209 @@
...
+                for i in (0..data_block_size as u64).step_by(FR_SIZE) {
+                    let input_index = (chunk_index as usize) + i as usize;
+                    let output_index = i as usize;
+
+                    // Get the `h` high bits from the node-index (sans partition-index).
+                    let node_index = input_index / FR_SIZE;
+                    let input_index_without_k = node_index & remove_k_from_node_index_mask;
+                    let high =
+                        input_index_without_k >> (node_index_bit_len - partition_bit_len - h);
+                    let rho: Fr = <TreeRHasher as Hasher>::Function::hash2(
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/vanilla.rs]: I believe that I've made the change, but a second review would be beneficial.

File: storage-proofs-update/src/vanilla.rs (hunk 4 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/vanilla.rs]: I believe that I've made the change, but a second review would be beneficial.

File: fil-proofs-param/Cargo.toml [hunk 12]
```diff
@@ -12,6 +12,7 @@ readme = "README.md"
 storage-proofs-core = { path = "../storage-proofs-core", version = "^10.0.0", default-features = false}
 storage-proofs-porep = { path = "../storage-proofs-porep", version = "^10.0.0", default-features = false }
 storage-proofs-post = { path = "../storage-proofs-post", version = "^10.0.0", default-features = false }
+storage-proofs-update = { path = "../storage-proofs-update", version = "^10.0.0", default-features = false }
```
Comment (dignifiedquire) [inline fil-proofs-param/Cargo.toml:15]: make sure to propagate features appropriately in all Cargo.toml files

File: fil-proofs-param/src/bin/paramcache.rs [hunk 13]
```diff
@@ -182,6 +212,30 @@ fn generate_params_porep(sector_size: u64, api_version: ApiVersion) {
                     .get(&sector_size)
                     .expect("unknown sector size"),
             ),
+            update_partitions: UpdateProofPartitions::from(partition_count(nodes_count)),
+            h_select: HSelect::from(hs(nodes_count)[0]),
```
Comment (dignifiedquire) [inline fil-proofs-param/src/bin/paramcache.rs:216]: This expression is used a lot, maybe add sth like  `HSelect::from_count(nodes_count)` as a method.

File: filecoin-proofs/src/types/porep_config.rs [hunk 14]
```diff
@@ -14,15 +14,20 @@ use storage_proofs_porep::stacked::{StackedCircuit, StackedCompound};
...
+        HSelect, PaddedBytesAmount, PoRepProofPartitions, SectorSize, UnpaddedBytesAmount,
+        UpdateProofPartitions,
+    },
 };
 
 #[derive(Clone, Copy, Debug)]
 pub struct PoRepConfig {
     pub sector_size: SectorSize,
     pub partitions: PoRepProofPartitions,
+    pub update_partitions: UpdateProofPartitions,
```
Comment (dignifiedquire) [inline filecoin-proofs/src/types/porep_config.rs:27]: Why is this added to the `PoRepConfig` (same for `h_select`)? It seems to me that there should be a new config type, rather than tacking this on the existing one, as they are somewhat separate

File: filecoin-proofs/src/api/mod.rs [hunk 15]
```diff
@@ -505,13 +510,24 @@ fn verify_store(config: &StoreConfig, arity: usize, required_configs: usize) ->
 
         let store_len = config.size.expect("disk store size not configured");
         for config in &configs {
+            let data_path = StoreConfig::data_path(&config.path, &config.id);
+            trace!(
+                "verify_store: {:?} has length {}",
```
Comment (dignifiedquire) [inline filecoin-proofs/src/api/mod.rs:515]: print the unit maybe?

File: filecoin-proofs/src/api/update.rs [hunk 16]
```diff
@@ -0,0 +1,616 @@
...
+    sector_key_cache_path: &Path,
+    replica_path: &Path,
+    replica_cache_path: &Path,
+    data_path: &Path,
+    comm_d_new: Commitment,
+) -> Result<()> {
+    info!("remove_data:start");
+
+    // NOTE: p_aux has comm_c and comm_r_last
+    let p_aux: PersistentAux<<Tree::Hasher as Hasher>::Domain> = {
```
Comment (dignifiedquire) [inline filecoin-proofs/src/api/update.rs:224]: This recovery of the different files is done quite often, maybe this could be extracted into a reusable function, sharing some code?

File: filecoin-proofs/src/api/update.rs [hunk 17]
```diff
@@ -0,0 +1,616 @@
...
+        <Tree::Hasher as Hasher>::Domain::try_from_bytes(&p_aux.comm_r_last.into_bytes())?,
+        u64::from(HSelect::from(porep_config)) as usize,
+    )?;
+
+    info!("remove_data:finish");
+    Ok(())
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn generate_single_partition_proof<Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
```
Comment (dignifiedquire) [inline filecoin-proofs/src/api/update.rs:251]: docs

File: filecoin-proofs/tests/api.rs [hunk 18]
```diff
@@ -1512,6 +1606,221 @@ fn create_seal_for_aggregation<R: Rng, Tree: 'static + MerkleTreeTrait>(
...
+fn create_seal_for_upgrade<R: Rng, Tree: 'static + MerkleTreeTrait<Hasher = TreeRHasher>>(
+    rng: &mut R,
+    sector_size: u64,
+    prover_id: ProverId,
+    porep_id: &[u8; 32],
+    api_version: ApiVersion,
+) -> Result<(SectorId, NamedTempFile, Commitment, TempDir)> {
+    init_logger();
+
+    let (mut piece_file, _piece_bytes) = generate_piece_file(sector_size)?;
```
Comment (dignifiedquire) [inline filecoin-proofs/tests/api.rs:1618]: I believe we have some specific functionality to create emtpy sectors, probably should also have a test where the seal is created using that.

File: filecoin-proofs/tests/api.rs [hunk 19]
```diff
@@ -1512,6 +1606,221 @@ fn create_seal_for_aggregation<R: Rng, Tree: 'static + MerkleTreeTrait>(
...
+    remove_encoded_data::<Tree>(
+        config,
+        remove_encoded_file.path(),
+        remove_encoded_cache_dir.path(),
+        new_sealed_sector_file.path(),
+        cache_dir.path(),
+        new_staged_sector_file.path(),
+        new_comm_d,
+    )?;
+    // When the data is removed, it MUST match the original sealed data.
```
Comment (dignifiedquire) [inline filecoin-proofs/tests/api.rs:1814]: Should check that the files are actually there & gone afterwards?

File: storage-proofs-core/src/gadgets/por.rs [hunk 20]
```diff
@@ -468,44 +466,125 @@ where
     Tree: MerkleTreeTrait,
     CS: ConstraintSystem<Fr>,
 {
-    let arity = Tree::Arity::to_usize();
-    let arity_bit_len = arity.trailing_zeros() as usize;
-    let challenge_bit_len = challenge_bits.len();
-    let height = path_values.len();
+    // This function assumes that `Tree`'s shape is valid, e.g. `base_arity > 0`, `if top_arity > 0
+    // then sub_arity > 0`, all arities are a power of two, etc., and that `path_values` corresponds
+    // to the tree arities.
```
Comment (dignifiedquire) [inline storage-proofs-core/src/gadgets/por.rs:471]: maybe add some `debug_assert`s to enforce this?

File: storage-proofs-update/src/challenges.rs [hunk 21]
```diff
@@ -0,0 +1,141 @@
+use blstrs::Scalar as Fr;
+use ff::{PrimeField, PrimeFieldBits};
+use filecoin_hashers::{poseidon::PoseidonHasher, HashFunction, Hasher};
+
+use crate::constants::{challenge_count, partition_count, TreeRDomain};
+
+pub struct Challenges {
```
Comment (dignifiedquire) [inline storage-proofs-update/src/challenges.rs:7]: docs

File: storage-proofs-update/src/challenges.rs [hunk 22]
```diff
@@ -0,0 +1,141 @@
...
+    use filecoin_hashers::Domain;
+    use rand::SeedableRng;
+    use rand_xorshift::XorShiftRng;
+    use storage_proofs_core::TEST_SEED;
+
+    use crate::constants::{TreeRDomain, ALLOWED_SECTOR_SIZES};
+
+    #[test]
+    fn test_challenges() {
+        let mut rng = XorShiftRng::from_seed(TEST_SEED);
```
Comment (dignifiedquire) [inline storage-proofs-update/src/challenges.rs:112]: Given this is a seeded rng, I would suggest to generate a small list of challenges, that are fixed, check them manually and then always compare against those in tests.

File: storage-proofs-update/src/circuit.rs [hunk 23]
```diff
@@ -0,0 +1,590 @@
...
+                bits_remaining -= sub_arity.trailing_zeros() as usize;
+            };
+
+            if top_arity > 0 {
+                sub_and_top_path.push(vec![None; top_arity - 1]);
+                bits_remaining -= top_arity.trailing_zeros() as usize;
+            };
+
+            let base_path_len = bits_remaining / base_arity.trailing_zeros() as usize;
+            let base_path = vec![vec![None; base_arity - 1]; base_path_len];
```
Comment (dignifiedquire) [inline storage-proofs-update/src/circuit.rs:161]: Why `base_arity - 1`? (probably obvious, I am just not seeing it right now)

File: storage-proofs-update/src/circuit.rs [hunk 24]
```diff
@@ -0,0 +1,590 @@
...
+
+        // Allocate the six least-significant bits of `h_select`.
+        let h_select_bits: Vec<AllocatedBit> =
+            allocated_num_to_allocated_bits(cs.namespace(|| "h_select_bits"), &h_select)?
+                .into_iter()
+                .take(h_select_bit_len)
+                .collect();
+
+        // phi = H(comm_d_new || comm_r_old)
+        let phi = <TreeR::Hasher as Hasher>::Function::hash2_circuit(
```
Comment (dignifiedquire) [inline storage-proofs-update/src/circuit.rs:335]: Should this really depend on the treeR hasher, or should this just be fixed to be poseidon?

File: storage-proofs-update/src/circuit.rs [hunk 25]
```diff
@@ -0,0 +1,590 @@
...
+            })
+            .collect::<Result<Vec<AllocatedNum<Fr>>, SynthesisError>>()?;
+
+        let partition_path = partition_path
+            .iter()
+            .enumerate()
+            .map(|(i, siblings)| {
+                AllocatedNum::alloc(
+                    cs.namespace(|| format!("partition_path_sibling_{}", i)),
+                    || siblings[0].ok_or(SynthesisError::AssignmentMissing),
```
Comment (dignifiedquire) [inline storage-proofs-update/src/circuit.rs:372]: shouldn't this be `siblings[i]`?

File: storage-proofs-update/src/compound.rs [hunk 26]
```diff
@@ -0,0 +1,147 @@
...
+
+        Ok(pub_inputs_circ.to_vec())
+    }
+
+    // Generates a partition's circuit. If the `k` argument is `Some` we overwrite `pub_inputs.k`
+    // with the `k` argument's value, otherwise if the `k` argument is `None` we use `pub_inputs.k`
+    // as the circuit's public-input.
+    fn circuit(
+        pub_inputs: &PublicInputs,
+        _priv_inputs: <EmptySectorUpdateCircuit<TreeR> as CircuitComponent>::ComponentPrivateInputs,
```
Comment (dignifiedquire) [inline storage-proofs-update/src/compound.rs:82]: this seems suspicious , why do we not need the `priv_inputs` at all?

File: storage-proofs-update/src/constants.rs [hunk 27]
```diff
@@ -0,0 +1,112 @@
+use filecoin_hashers::{
+    poseidon::{PoseidonDomain, PoseidonHasher},
+    sha256::{Sha256Domain, Sha256Hasher},
+};
+use generic_array::typenum::{Unsigned, U0, U2, U8};
+use storage_proofs_core::merkle::{BinaryMerkleTree, LCStore, LCTree, MerkleTreeTrait};
+
+// Sector-sizes measured in nodes.
+pub const SECTOR_SIZE_1_KIB: usize = 1 << 5;
```
Comment (dignifiedquire) [inline storage-proofs-update/src/constants.rs:9]: Can we share these constants with `filecoin-proofs`?

File: storage-proofs-update/src/gadgets.rs [hunk 28]
```diff
@@ -0,0 +1,467 @@
...
+        || "binary decomp",
+        |_| lc,
+        |lc| lc + CS::one(),
+        |lc| lc + num.get_variable(),
+    );
+
+    Ok(bits)
+}
+
+pub fn apex_por<CS: ConstraintSystem<Fr>>(
```
Comment (dignifiedquire) [inline storage-proofs-update/src/gadgets.rs:41]: docs please, this is a pretty complex thing

File: storage-proofs-update/src/vanilla.rs [hunk 29]
```diff
@@ -0,0 +1,1205 @@
...
+
+impl<'a, TreeR> ProofScheme<'a> for EmptySectorUpdate<TreeR>
+where
+    TreeR: 'static + MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    type SetupParams = SetupParams;
+    type PublicParams = PublicParams;
+    type PublicInputs = PublicInputs;
+    type PrivateInputs = PrivateInputs;
+    type Proof = PartitionProof<TreeR>;
```
Comment (dignifiedquire) [inline storage-proofs-update/src/vanilla.rs:239]: `PartitionProof` is a bit odd in naming here, could you explain the reasoning?

File: storage-proofs-update/src/vanilla.rs [hunk 30]
```diff
@@ -0,0 +1,1205 @@
...
+            partition_pub_inputs.k = k;
+            if !Self::verify(pub_params, &partition_pub_inputs, partition_proof)? {
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+
+    fn with_partition(mut pub_inputs: Self::PublicInputs, k: Option<usize>) -> Self::PublicInputs {
+        pub_inputs.k = k.unwrap();
```
Comment (dignifiedquire) [inline storage-proofs-update/src/vanilla.rs:725]: This shouldn't panic, or at least have a better error message

File: storage-proofs-update/src/vanilla.rs [hunk 31]
```diff
@@ -0,0 +1,1205 @@
...
+        let rhos: Vec<Fr> = (0..1 << h)
+            .map(|high| {
+                let high: TreeRDomain = Fr::from(high as u64).into();
+                <TreeRHasher as Hasher>::Function::hash2(&phi, &high).into()
+            })
+            .collect();
+
+        Vec::from_iter((0..end).step_by(data_block_size))
+            .into_par_iter()
+            .zip(new_replica_data.par_chunks_mut(data_block_size))
```
Comment (dignifiedquire) [inline storage-proofs-update/src/vanilla.rs:883]: is this a good parallelization strategy for 32/64gb sectors?

File: storage-proofs-update/src/vanilla.rs [hunk 32]
```diff
@@ -0,0 +1,1205 @@
...
+                    let high = node_index >> get_high_bits_shr;
+                    let rho = rhos[high];
+
+                    let sector_key_fr =
+                        bytes_into_fr(&sector_key_data[input_index..input_index + FR_SIZE])?;
+                    let staged_data_fr =
+                        bytes_into_fr(&staged_data[input_index..input_index + FR_SIZE])?;
+
+                    let new_replica_fr = sector_key_fr + (staged_data_fr * rho);
+                    let new_replica_bytes = fr_into_bytes(&new_replica_fr);
```
Comment (dignifiedquire) [inline storage-proofs-update/src/vanilla.rs:900]: could avoid this allocation by having a version of `fr_into_bytes` that writes directly into a slice

File: storage-proofs-update/src/circuit.rs [hunk 33]
```diff
@@ -0,0 +1,590 @@
...
+            &comm_r_old,
+        )?;
+
+        // Allocate private-inputs; excludes each challenge's Merkle proofs.
+
+        let comm_c = AllocatedNum::alloc(cs.namespace(|| "comm_c"), || {
+            comm_c.ok_or(SynthesisError::AssignmentMissing)
+        })?;
+
+        // TODO: rename `comm_r_last_old` and `comm_r_last_new` to `root_r_old` and `root_r_new`.
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/circuit.rs:452]: Yup, it will probably be the last thing that I do.

File: storage-proofs-update/src/compound.rs [hunk 34]
```diff
@@ -0,0 +1,147 @@
...
+    pub _tree_r: PhantomData<TreeR>,
+}
+
+impl<TreeR> CacheableParameters<EmptySectorUpdateCircuit<TreeR>, PublicParams>
+    for EmptySectorUpdateCompound<TreeR>
+where
+    TreeR: MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    fn cache_prefix() -> String {
+        format!("empty-sector-update-{}", TreeR::display())
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/compound.rs:30]: I believe this is covered by `<PublicParams as ParameterSetMetadata>::identifier()` (contains the sector-size) which is appended onto the `cache_prefix` path.

File: storage-proofs-update/src/compound.rs [hunk 35]
```diff
@@ -0,0 +1,147 @@
...
+where
+    TreeR: 'static + MerkleTreeTrait<Hasher = TreeRHasher>,
+{
+    // Generates a partition circuit's public-inputs. If the `k` argument is `Some` we overwrite
+    // `pub_inputs.k` with the `k` argument's value, otherwise if the `k` argument is `None` we use
+    // `pub_inputs.k` as the circuit's public-input.
+    fn generate_public_inputs(
+        pub_inputs: &PublicInputs,
+        pub_params: &PublicParams,
+        k: Option<usize>,
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/compound.rs:45]: It was done to accommodate the default methods in `CompoundProof` (`prove`, `verify`, etc.) which take the first partition's public-inputs as an argument and uses them to create each partition's `k in 0..partition_count` public-inputs via `generate_public_inputs(pub_inputs, pub_params, Some(k))`.

I think that this interface weirdness arises due to `generate_public_inputs` not expecting `k` to be in `PublicInputs`. I could change this by overwriting some of `CompoundProof`'s default methods.

File: storage-proofs-update/src/compound.rs [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+            comm_r_new,
+            h,
+            ..
+        } = *pub_inputs;
+
+        let h_select = hs(pub_params.sector_nodes)
+            .iter()
+            .position(|h_allowed| *h_allowed == h)
+            .map(|i| 1 << i as u64)
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/compound.rs]: Moving this to `circuit::PublicInputs::new()`.

File: storage-proofs-update/src/vanilla.rs (hunk 29 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/vanilla.rs:239]: I guess this is just preference, but I thought that it clarified the naming.

I chose the name to differentiate it from a `ChallengeProof` (which compose a partition proof) and from a `Proof`/`VanillaProof`, which (imho) is an ambiguous name that doesn't tell you whether the proof was generated for a single or multiple partition.
```rust
impl ProofScheme for EmptySectorUpdate {
    type Proof = PartitionProof;

    fn prove(...) -> Self::Proof;
}
```
makes it explicit that `prove` is called for each partition. Also, `PartitionProof` makes it clear that a single proof should be fed as the input to a single partition's circuit.

File: storage-proofs-update/src/gadgets.rs [hunk 37]
```diff
@@ -0,0 +1,467 @@
...
+    mut cs: CS,
+    apex_leafs: Vec<AllocatedNum<Fr>>,
+    partition_bits: Vec<AllocatedBit>,
+    partition_path: Vec<Vec<AllocatedNum<Fr>>>,
+    comm_d: AllocatedNum<Fr>,
+) -> Result<(), SynthesisError> {
+    // `apex_leafs.len()` is guaranteed to be a power of two.
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/gadgets.rs:48]: It is guaranteed because `constants::apex_leaf_count()` is always a power of two and because `apex_por` is not a public gadget; the circuit has already checked (before calling this gadget) that `apex_leafs.len() == apex_leaf_count(sector_size)`.

File: storage-proofs-update/src/circuit.rs [hunk 38]
```diff
@@ -0,0 +1,590 @@
...
+            let leaf_r_new = label_r_new(
+                cs.namespace(|| format!("leaf_r_new (c_index={})", c_index)),
+                &leaf_r_old,
+                &leaf_d_new,
+                &rho,
+            )?;
+
+            // Check that the calculated value of `leaf_r_new` agrees with the provided value.
+            if let Some(leaf_r_new) = leaf_r_new.get_value() {
+                assert_eq!(leaf_r_new, challenge_proof.leaf_r_new.unwrap());
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/circuit.rs:581]: So there is both, a vanilla sanity check and a constraint.

File: storage-proofs-update/src/compound.rs (hunk 26 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/compound.rs:75]: A partition's witness is fully derived from its `vanilla_proof`.

File: storage-proofs-update/src/compound.rs [hunk 39]
```diff
@@ -0,0 +1,147 @@
...
```
[the remark this answers is not in the document]
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/compound.rs]: Ok, moved that check into `circuit::PublicInputs::new()`.

File: storage-proofs-update/src/compound.rs (hunk 26 above)
Comment (dignifiedquire) [reply storage-proofs-update/src/compound.rs:75]: right

File: storage-proofs-update/src/constants.rs (hunk 27 above)
Comment (dignifiedquire) [reply storage-proofs-update/src/constants.rs:9]: if we need them in here, maybe export them and import them in filecoin-proofs, not sure

File: storage-proofs-update/src/gadgets.rs (hunk 37 above)
[the remark this answers is not in the document]
Comment (dignifiedquire) [reply storage-proofs-update/src/gadgets.rs:48]: I see

File: storage-proofs-update/src/vanilla.rs (hunk 29 above)
Comment (dignifiedquire) [reply storage-proofs-update/src/vanilla.rs:239]: Makes sense

File: storage-proofs-update/src/challenges.rs (hunk 22 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/challenges.rs:225]: Ok, I believe that I added this in [`test_challenges_against_hardcoded`](https://github.com/filecoin-project/rust-fil-proofs/blob/ce1836f6bc5dcaef79610727e433cacf7c8bf74c/storage-proofs-update/src/challenges.rs#L127).

File: storage-proofs-update/src/challenges.rs (hunk 21 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/challenges.rs:17]: Done

File: storage-proofs-update/src/circuit.rs (hunk 24 above)
Comment (DrPeterVanNostrand) [reply storage-proofs-update/src/circuit.rs]: I have hardcoded `phi` and `rho` to use Poseidon.

**Comment by cryptonemo** [comment]
Just merged the other GPU PR.  This PR is officially 'ready for review'.

File: storage-proofs-update/src/vanilla.rs [hunk 40]
```diff
@@ -850,43 +886,21 @@ where
...
-                i + 1,
-                tree_count
-            );
-            LCTree::<TreeRHasher, TreeR::Arity, U0, U0>::from_par_iter_with_config(
-                current_data,
-                config.clone(),
-            )?;
-        }
+        // This argument is currently unused by this invocation, but required for the API.
+        let mut unused_data = Data::empty();
```
Comment (dignifiedquire) [inline storage-proofs-update/src/vanilla.rs:894]: Maybe make the argument `Option` if it is not used in some cases?

Comment (cryptonemo) [reply storage-proofs-update/src/vanilla.rs:894]: I had tried that before, but ran into rust move issues :sweat_smile:

File: fil-proofs-tooling/Cargo.toml [hunk 41]
```diff
@@ -13,6 +13,7 @@ readme = "README.md"
 storage-proofs-core = { path = "../storage-proofs-core", version = "^10.0.0", default-features = false}
 storage-proofs-porep = { path = "../storage-proofs-porep", version = "^10.0.0", default-features = false }
 storage-proofs-post = { path = "../storage-proofs-post", version = "^10.0.0", default-features = false }
+storage-proofs-update = { path = "../storage-proofs-update", version = "^10.0.0", default-features = false }
```
Comment (qy3u) [inline fil-proofs-tooling/Cargo.toml:16]: `storage-proofs-update/cuda` and `storage-proofs-update/opencl` should be include in the features below.

Comment (cryptonemo) [reply fil-proofs-tooling/Cargo.toml:16]: Thanks, thought I had them all previously, but didn't.

File: storage-proofs-porep/src/stacked/vanilla/proof.rs [hunk 42]
```diff
@@ -804,28 +830,132 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
...
+                            .expect("failed to convert key");
+                    let encoded_node =
+                        encode::<<Tree::Hasher as Hasher>::Domain>(key_elem, data_node);
+                    data_node_bytes.copy_from_slice(AsRef::<[u8]>::as_ref(&encoded_node));
+
+                    encoded_node
+                })
+                .collect();
+
+            Ok(TreeRElementData::ElementList(encoded_data))
```
[the remark this answers is not in the document]
Comment (cryptonemo) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:897]: The code is already located in the helper method, we just have 2 versions of it depending on GPU/no-GPU support

[the remark this answers is not in the document]
Comment (qy3u) [reply storage-proofs-porep/src/stacked/vanilla/proof.rs:897]: Yes, there are 2 versions: GPU/non-GPU.But there are 3 pieces of code:
1. GPU feature  enabled & `use_gpu_tree_builder` was set.
2. GPU feature  enabled & `use_gpu_tree_builder` was not set
3. GPU feature not enabled.

As I can see, the code for  `2` and `3` are exactly the same in the current implementation(sames like just copy-paste).What I mean is that may be we can merge them into one helper funtion. Just an advice.

File: storage-proofs-update/src/vanilla.rs [hunk 43]
```diff
@@ -0,0 +1,1172 @@
...
+pub struct PrivateInputs {
+    pub comm_c: TreeRDomain,
+    pub tree_r_old_config: StoreConfig,
+    // Path to old replica.
+    pub old_replica_path: PathBuf,
+    pub tree_d_new_config: StoreConfig,
+    pub tree_r_new_config: StoreConfig,
+    // Path to new replica.
+    pub replica_path: PathBuf,
+}
```
Comment (vmx) [inline storage-proofs-update/src/vanilla.rs:142]: Do all members (also for the `PublicParams` above) need to be public? I try to keep as much out of the public API as possible. Once somrthing is in the public API, it's tricky to get it out again. E.g. If you want to change a type (let's say you change from `PathBuf` to `Path`) and it's public, it's a breaking change and you have to be very careful about it.

**Review by vmx — COMMENTED** [review]
I've done a general Rust review, I didn't really get into the algorithms.
